From a143ce2a68250c5475af3fc6a35019c3599ff930 Mon Sep 17 00:00:00 2001
From: Thomas Colliers <mail@thomascolliers.com>
Date: Sun, 16 Aug 2026 20:30:35 +0200
Subject: [PATCH 1/5] media: dt-bindings: camss: Add qcom,milos-camss

Document the camera subsystem on SM7635 (Milos), comprising four
CSIPHY, three CSID and three TFE instances.

All four CSIPHYs are fed by the same two rails, so document
vdd-csiphy-0p9-supply and vdd-csiphy-1p2-supply as required
properties.

Both are named after their voltage rather than after a phy/pll split,
following qcom,x1e80100-camss.yaml. The downstream device tree
identifies the rails by voltage only, so which of the two feeds the
PLL is not known here, and a vdda-phy/vdda-pll spelling would encode
that guess into the binding.

Assisted-by: pi:claude-opus-5
Suggested-by: Jorijn van der Graaf <jorijnvdgraaf@catcrafts.net>
Signed-off-by: Thomas Colliers <mail@thomascolliers.com>
---
 .../bindings/media/qcom,milos-camss.yaml      | 291 ++++++++++++++++++
 1 file changed, 291 insertions(+)
 create mode 100644 Documentation/devicetree/bindings/media/qcom,milos-camss.yaml

diff --git a/Documentation/devicetree/bindings/media/qcom,milos-camss.yaml b/Documentation/devicetree/bindings/media/qcom,milos-camss.yaml
new file mode 100644
index 000000000000..fc9331cecb29
--- /dev/null
+++ b/Documentation/devicetree/bindings/media/qcom,milos-camss.yaml
@@ -0,0 +1,291 @@
+# SPDX-License-Identifier: (GPL-2.0-only OR BSD-2-Clause)
+%YAML 1.2
+---
+$id: http://devicetree.org/schemas/media/qcom,milos-camss.yaml#
+$schema: http://devicetree.org/meta-schemas/core.yaml#
+
+title: Qualcomm Milos Camera Subsystem (CAMSS)
+
+maintainers:
+  - Thomas Colliers <mail@thomascolliers.com>
+
+description:
+  The CAMSS on Milos comprises four CSI PHYs, three CSI decoders and three
+  real-time image processing engines.
+
+properties:
+  compatible:
+    const: qcom,milos-camss
+
+  reg:
+    items:
+      - description: CAMSS top registers
+      - description: CSID 0 registers
+      - description: CSID 1 registers
+      - description: CSID 2 registers
+      - description: CSIPHY 0 registers
+      - description: CSIPHY 1 registers
+      - description: CSIPHY 2 registers
+      - description: CSIPHY 3 registers
+      - description: TFE 0 registers
+      - description: TFE 1 registers
+      - description: TFE 2 registers
+
+  reg-names:
+    items:
+      - const: top
+      - const: csid0
+      - const: csid1
+      - const: csid2
+      - const: csiphy0
+      - const: csiphy1
+      - const: csiphy2
+      - const: csiphy3
+      - const: vfe0
+      - const: vfe1
+      - const: vfe2
+
+  clocks:
+    maxItems: 26
+
+  clock-names:
+    items:
+      - const: cpas_ahb
+      - const: soc_ahb
+      - const: csiphy0
+      - const: csiphy0_timer
+      - const: csiphy1
+      - const: csiphy1_timer
+      - const: csiphy2
+      - const: csiphy2_timer
+      - const: csiphy3
+      - const: csiphy3_timer
+      - const: tfe0_ahb
+      - const: tfe1_ahb
+      - const: tfe2_ahb
+      - const: csi0
+      - const: csi1
+      - const: csi2
+      - const: vfe0_cphy_rx
+      - const: vfe1_cphy_rx
+      - const: vfe2_cphy_rx
+      - const: vfe0
+      - const: vfe1
+      - const: vfe2
+      - const: camnoc_rt_axi
+      - const: camnoc_nrt_axi
+      - const: camnoc_axi_hf
+      - const: camnoc_axi_sf
+
+  interrupts:
+    maxItems: 10
+
+  interrupt-names:
+    items:
+      - const: csid0
+      - const: csid1
+      - const: csid2
+      - const: csiphy0
+      - const: csiphy1
+      - const: csiphy2
+      - const: csiphy3
+      - const: vfe0
+      - const: vfe1
+      - const: vfe2
+
+  interconnects:
+    items:
+      - description: AHB path
+      - description: High-frequency memory path
+      - description: System-frequency memory path
+
+  interconnect-names:
+    items:
+      - const: ahb
+      - const: hf_mnoc
+      - const: sf_mnoc
+
+  iommus:
+    maxItems: 2
+
+  power-domains:
+    items:
+      - description: CAMSS top GDSC
+
+  vdd-csiphy-0p9-supply:
+    description:
+      0.9V supply to a PHY.
+
+  vdd-csiphy-1p2-supply:
+    description:
+      1.2V supply to a PHY.
+
+  ports:
+    $ref: /schemas/graph.yaml#/properties/ports
+
+    description: CSI input ports.
+
+    patternProperties:
+      "^port@[0-3]$":
+        $ref: /schemas/graph.yaml#/$defs/port-base
+        unevaluatedProperties: false
+
+        description: Input port for receiving CSI data from a CSIPHY.
+
+        properties:
+          endpoint:
+            $ref: video-interfaces.yaml#
+            unevaluatedProperties: false
+
+            properties:
+              data-lanes:
+                minItems: 1
+                maxItems: 4
+
+              bus-type:
+                enum:
+                  - 1 # MEDIA_BUS_TYPE_CSI2_CPHY
+                  - 4 # MEDIA_BUS_TYPE_CSI2_DPHY
+
+            required:
+              - data-lanes
+
+required:
+  - compatible
+  - reg
+  - reg-names
+  - clocks
+  - clock-names
+  - interrupts
+  - interrupt-names
+  - interconnects
+  - interconnect-names
+  - iommus
+  - power-domains
+  - vdd-csiphy-0p9-supply
+  - vdd-csiphy-1p2-supply
+  - ports
+
+additionalProperties: false
+
+examples:
+  - |
+    #include <dt-bindings/clock/qcom,milos-camcc.h>
+    #include <dt-bindings/interconnect/qcom,icc.h>
+    #include <dt-bindings/interconnect/qcom,milos-rpmh.h>
+    #include <dt-bindings/interrupt-controller/arm-gic.h>
+
+    soc {
+        #address-cells = <2>;
+        #size-cells = <2>;
+
+        camss: camss@ac13000 {
+            compatible = "qcom,milos-camss";
+
+            reg = <0x0 0x0ac13000 0x0 0x1000>,
+                  <0x0 0x0ac62000 0x0 0x1000>,
+                  <0x0 0x0ac71000 0x0 0x1000>,
+                  <0x0 0x0ac80000 0x0 0x1000>,
+                  <0x0 0x0ace4000 0x0 0x2000>,
+                  <0x0 0x0ace6000 0x0 0x2000>,
+                  <0x0 0x0ace8000 0x0 0x2000>,
+                  <0x0 0x0acea000 0x0 0x2000>,
+                  <0x0 0x0ac63800 0x0 0xb800>,
+                  <0x0 0x0ac72800 0x0 0xb800>,
+                  <0x0 0x0ac81800 0x0 0xb800>;
+            reg-names = "top",
+                        "csid0", "csid1", "csid2",
+                        "csiphy0", "csiphy1", "csiphy2", "csiphy3",
+                        "vfe0", "vfe1", "vfe2";
+
+            clocks = <&camcc CAM_CC_CPAS_AHB_CLK>,
+                     <&camcc CAM_CC_SOC_AHB_CLK>,
+                     <&camcc CAM_CC_CSIPHY0_CLK>,
+                     <&camcc CAM_CC_CSI0PHYTIMER_CLK>,
+                     <&camcc CAM_CC_CSIPHY1_CLK>,
+                     <&camcc CAM_CC_CSI1PHYTIMER_CLK>,
+                     <&camcc CAM_CC_CSIPHY2_CLK>,
+                     <&camcc CAM_CC_CSI2PHYTIMER_CLK>,
+                     <&camcc CAM_CC_CSIPHY3_CLK>,
+                     <&camcc CAM_CC_CSI3PHYTIMER_CLK>,
+                     <&camcc CAM_CC_TFE_0_AHB_CLK>,
+                     <&camcc CAM_CC_TFE_1_AHB_CLK>,
+                     <&camcc CAM_CC_TFE_2_AHB_CLK>,
+                     <&camcc CAM_CC_TFE_0_CSID_CLK>,
+                     <&camcc CAM_CC_TFE_1_CSID_CLK>,
+                     <&camcc CAM_CC_TFE_2_CSID_CLK>,
+                     <&camcc CAM_CC_TFE_0_CPHY_RX_CLK>,
+                     <&camcc CAM_CC_TFE_1_CPHY_RX_CLK>,
+                     <&camcc CAM_CC_TFE_2_CPHY_RX_CLK>,
+                     <&camcc CAM_CC_TFE_0_CLK>,
+                     <&camcc CAM_CC_TFE_1_CLK>,
+                     <&camcc CAM_CC_TFE_2_CLK>,
+                     <&camcc CAM_CC_CAMNOC_RT_AXI_CLK>,
+                     <&camcc CAM_CC_CAMNOC_NRT_AXI_CLK>,
+                     <&camcc CAM_CC_CAMNOC_AXI_HF_CLK>,
+                     <&camcc CAM_CC_CAMNOC_AXI_SF_CLK>;
+            clock-names = "cpas_ahb", "soc_ahb",
+                          "csiphy0", "csiphy0_timer",
+                          "csiphy1", "csiphy1_timer",
+                          "csiphy2", "csiphy2_timer",
+                          "csiphy3", "csiphy3_timer",
+                          "tfe0_ahb", "tfe1_ahb", "tfe2_ahb",
+                          "csi0", "csi1", "csi2",
+                          "vfe0_cphy_rx", "vfe1_cphy_rx", "vfe2_cphy_rx",
+                          "vfe0", "vfe1", "vfe2",
+                          "camnoc_rt_axi", "camnoc_nrt_axi",
+                          "camnoc_axi_hf", "camnoc_axi_sf";
+
+            interrupts = <GIC_SPI 464 IRQ_TYPE_EDGE_RISING>,
+                         <GIC_SPI 466 IRQ_TYPE_EDGE_RISING>,
+                         <GIC_SPI 203 IRQ_TYPE_EDGE_RISING>,
+                         <GIC_SPI 477 IRQ_TYPE_EDGE_RISING>,
+                         <GIC_SPI 478 IRQ_TYPE_EDGE_RISING>,
+                         <GIC_SPI 479 IRQ_TYPE_EDGE_RISING>,
+                         <GIC_SPI 448 IRQ_TYPE_EDGE_RISING>,
+                         <GIC_SPI 602 IRQ_TYPE_EDGE_RISING>,
+                         <GIC_SPI 604 IRQ_TYPE_EDGE_RISING>,
+                         <GIC_SPI 688 IRQ_TYPE_EDGE_RISING>;
+            interrupt-names = "csid0", "csid1", "csid2",
+                              "csiphy0", "csiphy1", "csiphy2", "csiphy3",
+                              "vfe0", "vfe1", "vfe2";
+
+            interconnects = <&gem_noc MASTER_APPSS_PROC QCOM_ICC_TAG_ACTIVE_ONLY
+                             &cnoc_cfg SLAVE_CAMERA_CFG QCOM_ICC_TAG_ACTIVE_ONLY>,
+                            <&mmss_noc MASTER_CAMNOC_HF QCOM_ICC_TAG_ALWAYS
+                             &mc_virt SLAVE_EBI1 QCOM_ICC_TAG_ALWAYS>,
+                            <&mmss_noc MASTER_CAMNOC_SF QCOM_ICC_TAG_ALWAYS
+                             &mc_virt SLAVE_EBI1 QCOM_ICC_TAG_ALWAYS>;
+            interconnect-names = "ahb", "hf_mnoc", "sf_mnoc";
+
+            iommus = <&apps_smmu 0x0880 0x20>,
+                     <&apps_smmu 0x08a0 0x20>;
+
+            power-domains = <&camcc CAM_CC_CAMSS_TOP_GDSC>;
+
+            vdd-csiphy-0p9-supply = <&vreg_l2b>;
+            vdd-csiphy-1p2-supply = <&vreg_l4b>;
+
+            ports {
+                #address-cells = <1>;
+                #size-cells = <0>;
+
+                port@0 {
+                    reg = <0>;
+                };
+
+                port@1 {
+                    reg = <1>;
+                };
+
+                port@2 {
+                    reg = <2>;
+                };
+
+                port@3 {
+                    reg = <3>;
+                };
+            };
+        };
+    };
+...
